src/liballoc/vec.rs

```diff
@@ -1306,6 +1306,48 @@ impl<T> Vec<T> {
         }
         other
     }
+
+    /// Resizes the `Vec` in-place so that `len` is equal to `new_len`.
+    ///
+    /// If `new_len` is greater than `len`, the `Vec` is extended by the
+    /// difference, with each additional slot filled with the result of
+    /// closure `f`. If `new_len` is less than `len`, the `Vec` is simply
+    /// truncated.
+    ///
+    /// This method uses a closure to create new values on every push. If
+    /// you'd rather [`Clone`] a given value, use [`resize`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(vec_resize_with)]
+    ///
+    /// let mut vec = vec![1, 2, 3];
+    /// vec.resize_with(5, Default::default);
+    /// assert_eq!(vec, [1, 2, 3, 0, 0]);
+    ///
+    /// let mut vec = vec![1, 2, 3, 4];
+    /// vec.resize_with(2, Default::default);
+    /// assert_eq!(vec, [1, 2]);
+    /// ```
+    ///
+    /// [`resize`]: #method.resize
+    /// [`Clone`]: ../../std/clone/trait.Clone.html
+    #[unstable(feature = "vec_resize_with", issue = "41758")]
+    pub fn resize_with<F>(&mut self, new_len: usize, mut f: F)
+        where F: FnMut() -> T
+    {
+        let len = self.len();
```

Comment:
Should the be implemented using extend_with, like the other two are?  Something like

```rust
struct ExtendFunc<F>(F);
impl<T, F: FnMut() -> T> ExtendWith<T> for ExtendFunct<F> {
    fn next(&self) -> T { (self.0)() }
    fn last(self) -> T { (self.0)() }
}
```

(Would it ever be possible for a closure to implement its `FnOnce` differently from its `FnMut`, to take advantage of the difference here the way `ExtendElement` does?)

Reply:
As I understand it, the value returned by `f` is already moved into place for each iteration in the current implementation of `resize_with()`, so it seems like the `ExtendWith` mechanism doesn't make sense here.

Reply:
I believe this should use `ExtendWith`: `ExtendWith` is not only used to generalize and avoid the last clone, `extend_with` is also implemented in a way that makes optimizations more likely, as I understand the code.

Reply:
Okay, coming up.

Reply:
Hmm, I think that runs into mutability problems. `ExtendWith::next()` is defined as taking `&self`, in which case the mutability of the `FnMut` seems inaccessible through implementing `ExtendFunc` the way @scottmcm proposes. I don't think I see a way out of that...

Reply:
Good point!  ExtendWith is non-pub, though, so maybe just change it to take `&mut self`?  It doesn't look like an instance of it ever needs to be shared...

Reply:
That works, I guess.